File: OJ practices/LeetCode题目集/86. Partition List(medium).java
/*
 * @Author: SourDumplings
 * @Date: 2020-08-28 10:38:46
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/partition-list/
 */

class Solution
{
    public ListNode partition(ListNode head, int x)
    {
        ListNode before = new ListNode(0);
        ListNode after = new ListNode(0);
        ListNode p = head;
        ListNode pB = before;
        ListNode pA = after;
        while (p != null)
        {
            if (p.val < x)
            {
                pB.next = new ListNode(p.val);
                pB = pB.next;
            }
            else
            {
                pA.next = new ListNode(p.val);
                pA = pA.next;
            }
            p = p.next;
        }
        pB.next = after.next;
        return before.next;
    }
}

File: OJ practices/LeetCode题目集/86. Partition List(medium).cpp
/**
 * @file 86. Partition List(medium).cpp
 * @author SourDumplings (sourdumplings@qq.com)
 * @brief https://leetcode.cn/problems/partition-list/?envType=study-plan-v2&envId=top-interview-150
 * @version 1.0.0
 * @date 2024-11-30
 * 
 * @copyright Copyright (c) 2024 SourDumplings
 * 
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
	ListNode *partition(ListNode *head, int x) {
		ListNode *tempHead1 = new ListNode(0, head);
		ListNode *f1 = tempHead1;
		ListNode *tempHead2 = new ListNode(0);
		ListNode *f2 = tempHead2;
		while (f1->next)
		{
			ListNode *p = f1->next;
			if (p->val < x)
			{
				f1->next = p->next;
				p->next = nullptr;
				f2->next = p;
				f2 = p;
			}
			else
			{
				f1 = f1->next;
			}
		}
		f2->next = tempHead1->next;
		head = tempHead2->next;
		delete tempHead1;
		delete tempHead2;
		return head;
	}
};
